test/e2e/apps/statefulset.go

```diff
@@ -1595,19 +1601,32 @@ var _ = SIGDescribe("StatefulSet", func() {
 
 			ginkgo.By("Confirming 2 replicas, with start ordinal 3")
 			pods := e2estatefulset.GetPodList(ctx, c, ss)
-			expectPodNames(pods, []string{"ss-3", "ss-4"})
+			err = expectPodNames(pods, []string{"ss-3", "ss-4"})
+			framework.ExpectNoError(err)
 
 			ginkgo.By("Removing .spec.ordinals")
 			ss, err = updateStatefulSetWithRetries(ctx, c, ns, ss.Name, func(update *appsv1.StatefulSet) {
 				update.Spec.Ordinals = nil
 			})
 			framework.ExpectNoError(err)
+
+			// since we are replacing 2 pods for 2, we need to ensure we wait
+			// for the new ones to show up, not just for any random 2
+			framework.Logf("Confirming 2 replicas, with start ordinal 0")
+			e2estatefulset.WaitForState(ctx, c, ss,
+				func(intSet *appsv1.StatefulSet, pl *v1.PodList) (bool, error) {
+					if intSet.Status.ObservedGeneration < ss.Generation {
+						return false, nil
+					}
+					if err := expectPodNames(pl, []string{"ss-0", "ss-1", "ss-2"}); err != nil {
```

Comment:
The fix looks good to me, but why do we wait for 3 pods? Aren't we expecting two replicas? `"ss-0"` and `"ss-1"`? It might skip this state since the poll period is 10s.